File: src/engine/backend-cpu.js
// CPU reference backend.
//
// Every op the model needs is implemented here in plain JS. It is the ground
// truth the WebGPU backend is checked against, and it lets the whole transformer
// run (slowly) with no GPU — which is how the Node test-suite validates the math.
//
// Tensors are { data: Float32Array, shape: number[] }.

export class CpuTensor {
  constructor(data, shape) {
    this.data = data;
    this.shape = shape;
  }
  get size() {
    return this.data.length;
  }
}

const t = (data, shape) => new CpuTensor(data, shape);
const numel = (shape) => shape.reduce((a, b) => a * b, 1);

export class CpuBackend {
  constructor() {
    this.name = "cpu";
  }

  async init() {}
  async readback(x) {
    return x.data;
  }
  destroy() {}

  tensor(data, shape) {
    return t(data instanceof Float32Array ? data : Float32Array.from(data), shape);
  }
  zeros(shape) {
    return t(new Float32Array(numel(shape)), shape);
  }

  // rows[i] = table[ids[i]] * scale ; table is [V, D] (f32 or quantized).
  embedRows(table, ids, scale = 1.0) {
    const D = table.shape[1];
    const out = new Float32Array(ids.length * D);
    if (table.quant) {
      for (let i = 0; i < ids.length; i++)
        for (let d = 0; d < D; d++) out[i * D + d] = this._deq(table.quant, ids[i], d) * scale;
      return t(out, [ids.length, D]);
    }
    for (let i = 0; i < ids.length; i++) {
      const src = ids[i] * D;
      for (let d = 0; d < D; d++) out[i * D + d] = table.data[src + d] * scale;
    }
    return t(out, [ids.length, D]);
  }

  // RMSNorm: y = x/rms(x) * weight. The qat-mobile checkpoint stores norm
  // weights as the FULL multiplier (the reference runtime applies plain `w`,
  // not HF-Gemma's `1 + w`). Pass weight = null for an unweighted norm
  // (used on v ahead of attention).
  rmsNorm(x, weight, eps) {
    const [T, D] = x.shape;
    const out = new Float32Array(T * D);
    for (let i = 0; i < T; i++) {
      let ss = 0;
      const base = i * D;
      for (let d = 0; d < D; d++) {
        const v = x.data[base + d];
        ss += v * v;
      }
      const inv = 1 / Math.sqrt(ss / D + eps);
      for (let d = 0; d < D; d++) {
        out[base + d] = x.data[base + d] * inv * (weight ? weight.data[d] : 1);
      }
    }
    return t(out, [T, D]);
  }

  // A quantized weight kept un-expanded. `q` = { qbytes:Uint8Array, scales:
  // Float32Array, shape:[N,K], bits, signedI8, groupSize }.
  quantTensor(q) {
    const perByte = q.signedI8 ? 1 : 8 / q.bits;
    return { quant: { ...q, perByte }, shape: q.shape, size: q.shape[0] * q.shape[1] };
  }

  // Dequantize a single weight element matching the WGSL unpack exactly.
  _deq(q, n, k) {
    const { qbytes, scales, shape, bits, signedI8, groupSize, perByte } = q;
    const K = shape[1];
    const rowBytes = signedI8 ? K : K / perByte;
    let qv;
    if (signedI8) {
      qv = (qbytes[n * rowBytes + k] << 24) >> 24;
    } else {
      const b = qbytes[n * rowBytes + ((k / perByte) | 0)];
      const shift = (k % perByte) * bits;
      // offset binary: value = raw - zeroPoint, zeroPoint = 2^(bits-1)
      qv = ((b >> shift) & ((1 << bits) - 1)) - (1 << (bits - 1));
    }
    const sStride = Math.ceil(K / groupSize);
    return qv * scales[n * sStride + ((k / groupSize) | 0)];
  }

  // y = x @ W^T ; x is [T, K], W is [N, K] (row-major like PyTorch nn.Linear).
  linear(x, W) {
    const [T, K] = x.shape;
    const N = W.shape[0];
    const out = new Float32Array(T * N);
    if (W.quant) {
      for (let i = 0; i < T; i++)
        for (let n = 0; n < N; n++) {
          let acc = 0;
          for (let k = 0; k < K; k++) acc += x.data[i * K + k] * this._deq(W.quant, n, k);
          out[i * N + n] = acc;
        }
      return t(out, [T, N]);
    }
    for (let i = 0; i < T; i++) {
      const xb = i * K;
      for (let n = 0; n < N; n++) {
        const wb = n * K;
        let acc = 0;
        for (let k = 0; k < K; k++) acc += x.data[xb + k] * W.data[wb + k];
        out[i * N + n] = acc;
      }
    }
    return t(out, [T, N]);
  }

  add(a, b) {
    const out = new Float32Array(a.size);
    for (let i = 0; i < out.length; i++) out[i] = a.data[i] + b.data[i];
    return t(out, a.shape.slice());
  }
  mul(a, b) {
    const out = new Float32Array(a.size);
    for (let i = 0; i < out.length; i++) out[i] = a.data[i] * b.data[i];
    return t(out, a.shape.slice());
  }
  scale(a, s) {
    const out = new Float32Array(a.size);
    for (let i = 0; i < out.length; i++) out[i] = a.data[i] * s;
    return t(out, a.shape.slice());
  }
  // Multiply by a learned scalar that lives in a [1] tensor (layer_scalar).
  scaleByTensor(a, s) {
    return this.scale(a, s.data[0]);
  }

  geluTanh(x) {
    const out = new Float32Array(x.size);
    const c = Math.sqrt(2 / Math.PI);
    for (let i = 0; i < out.length; i++) {
      const v = x.data[i];
      out[i] = 0.5 * v * (1 + Math.tanh(c * (v + 0.044715 * v * v * v)));
    }
    return t(out, x.shape.slice());
  }

  // GeGLU: gelu(gate) * up. gate/up are [T, I].
  geglu(gate, up) {
    return this.mul(this.geluTanh(gate), up);
  }

  // cap * tanh(x / cap), elementwise (logit / attn soft-capping).
  softcap(x, cap) {
    const out = new Float32Array(x.size);
    for (let i = 0; i < out.length; i++) out[i] = cap * Math.tanh(x.data[i] / cap);
    return t(out, x.shape.slice());
  }

  // RoPE on q/k laid out [T, heads, headDim]. Rotates only the first
  // `rotaryDim` dims (partial rotary). Neighbour-pair (interleaved) convention
  // is avoided in favour of the half-split convention used by HF Gemma.
  rope(x, positions, theta, headDim, heads, rotaryDim) {
    const [T] = x.shape;
    const out = Float32Array.from(x.data);
    const half = rotaryDim >> 1;
    for (let i = 0; i < T; i++) {
      const pos = positions[i];
      for (let h = 0; h < heads; h++) {
        const base = (i * heads + h) * headDim;
        for (let j = 0; j < half; j++) {
          const freq = Math.pow(theta, -(2 * j) / rotaryDim);
          const ang = pos * freq;
          const cos = Math.cos(ang), sin = Math.sin(ang);
          const a = x.data[base + j];
          const b = x.data[base + j + half];
          out[base + j] = a * cos - b * sin;
          out[base + j + half] = b * cos + a * sin;
        }
      }
    }
    return t(out, x.shape.slice());
  }

  // Multi/grouped-query causal attention with optional sliding window.
  // q:[T,Hq,Dh] k:[S,Hkv,Dh] v:[S,Hkv,Dv]; positions length T and S (kv).
  // Returns [T, Hq*Dv].
  attention(q, k, v, { qPos, kPos, scale, slidingWindow, attnSoftcap }) {
    const [T, Hq, Dh] = q.shape;
    const S = k.shape[0];
    const Hkv = k.shape[1];
    const Dv = v.shape[2];
    const group = Hq / Hkv;
    const out = new Float32Array(T * Hq * Dv);
    const scores = new Float32Array(S);
    for (let i = 0; i < T; i++) {
      const qpos = qPos[i];
      for (let h = 0; h < Hq; h++) {
        const kvh = Math.floor(h / group);
        const qb = (i * Hq + h) * Dh;
        let maxv = -Infinity;
        for (let j = 0; j < S; j++) {
          const kpos = kPos[j];
          // causal + optional sliding-window mask
          if (kpos > qpos || (slidingWindow > 0 && qpos - kpos >= slidingWindow)) {
            scores[j] = -Infinity;
            continue;
          }
          const kb = (j * Hkv + kvh) * Dh;
          let dot = 0;
          for (let d = 0; d < Dh; d++) dot += q.data[qb + d] * k.data[kb + d];
          dot *= scale;
          if (attnSoftcap > 0) dot = attnSoftcap * Math.tanh(dot / attnSoftcap);
          scores[j] = dot;
          if (dot > maxv) maxv = dot;
        }
        // softmax over valid keys
        let sum = 0;
        for (let j = 0; j < S; j++) {
          if (scores[j] === -Infinity) {
            scores[j] = 0;
          } else {
            const e = Math.exp(scores[j] - maxv);
            scores[j] = e;
            sum += e;
          }
        }
        const inv = sum > 0 ? 1 / sum : 0;
        const ob = (i * Hq + h) * Dv;
        for (let j = 0; j < S; j++) {
          const w = scores[j] * inv;
          if (w === 0) continue;
          const vb = (j * Hkv + kvh) * Dv;
          for (let d = 0; d < Dv; d++) out[ob + d] += w * v.data[vb + d];
        }
      }
    }
    return t(out, [T, Hq * Dv]);
  }

  // Reshape view (no copy needed for CPU).
  reshape(x, shape) {
    return t(x.data, shape);
  }

  // Contiguous row slice: rows [start, start+count) of a [rows, dim] tensor.
  sliceRows(x, start, count) {
    const dim = x.shape[1];
    return t(x.data.slice(start * dim, (start + count) * dim), [count, dim]);
  }

  // Column slice: [:, offset:offset+width] of a [rows, stride] tensor.
  sliceCols(x, offset, width) {
    const [rows, stride] = x.shape;
    const out = new Float32Array(rows * width);
    for (let r = 0; r < rows; r++) {
      out.set(x.data.subarray(r * stride + offset, r * stride + offset + width), r * width);
    }
    return t(out, [rows, width]);
  }

  // Concatenate along dim 0 (sequence). Both must share trailing dims.
  concatRows(a, b) {
    if (!a) return t(Float32Array.from(b.data), b.shape.slice());
    const out = new Float32Array(a.size + b.size);
    out.set(a.data, 0);
    out.set(b.data, a.size);
    return t(out, [a.shape[0] + b.shape[0], ...a.shape.slice(1)]);
  }

  // KV-cache append (same contract as WebGpuBackend.kvAppend). The CPU
  // reference just concatenates; `dead` mirrors the GPU retirement signal but
  // is ignored because CpuBackend has no freeTensor.
  kvAppend(existing, add) {
    return { tensor: this.concatRows(existing, add), dead: existing };
  }
}

File: rust/src/backend_cpu.rs
//! CPU reference backend.
//!
//! Every op the model needs is implemented here in plain Rust. It is the
//! ground truth the wgpu backend is checked against, and it lets the whole
//! transformer run (slowly) with no GPU — which is how the test-suite
//! validates the math.

use std::sync::Arc;

use crate::backend::{AttnOpts, Backend, KvAppend, QuantWeight, TensorShape};

/// A CPU tensor: f32 data + shape, or a packed quantized weight.
#[derive(Clone, Debug)]
pub struct CpuTensor {
    pub data: Arc<Vec<f32>>,
    pub shape: Vec<usize>,
    pub quant: Option<Arc<QuantWeight>>,
}

impl CpuTensor {
    pub fn new(data: Vec<f32>, shape: &[usize]) -> Self {
        Self { data: Arc::new(data), shape: shape.to_vec(), quant: None }
    }
    pub fn data(&self) -> &[f32] {
        &self.data
    }
}

impl TensorShape for CpuTensor {
    fn shape(&self) -> &[usize] {
        &self.shape
    }
    fn size(&self) -> usize {
        if self.quant.is_some() { self.shape.iter().product() } else { self.data.len() }
    }
}

fn t(data: Vec<f32>, shape: &[usize]) -> CpuTensor {
    CpuTensor::new(data, shape)
}

/// A raw pointer that may cross into scoped threads; the callers partition the
/// pointee into disjoint row ranges, one per thread.
#[derive(Clone, Copy)]
struct SyncPtr(*mut f32);
unsafe impl Send for SyncPtr {}
unsafe impl Sync for SyncPtr {}
impl SyncPtr {
    // accessed through a method so closures capture the wrapper, not the raw field
    fn get(&self) -> *mut f32 {
        self.0
    }
}

pub fn gelu_tanh_scalar(v: f32) -> f32 {
    let c = (2.0f32 / std::f32::consts::PI).sqrt();
    0.5 * v * (1.0 + (c * (v + 0.044715 * v * v * v)).tanh())
}

#[derive(Default)]
pub struct CpuBackend;

/// Below this many multiply-adds an op runs on the calling thread.
const PAR_MIN_WORK: usize = 1 << 14;

/// Run `f(start, end)` over `[0, n)` split into contiguous chunks across the
/// available cores (scoped threads, no pool). Each chunk owns disjoint output
/// rows, so the per-row arithmetic — and therefore every result — is identical
/// to the sequential loop.
fn par_ranges(n: usize, work: usize, f: impl Fn(usize, usize) + Sync) {
    // GEMMA_CPU_THREADS overrides the core count (1 = sequential reference).
    let threads = std::env::var("GEMMA_CPU_THREADS")
        .ok()
        .and_then(|v| v.parse::<usize>().ok())
        .filter(|&t| t >= 1)
        .unwrap_or_else(|| std::thread::available_parallelism().map(|x| x.get()).unwrap_or(1))
        .min(n)
        .max(1);
    if threads == 1 || work < PAR_MIN_WORK {
        f(0, n);
        return;
    }
    let chunk = n.div_ceil(threads);
    std::thread::scope(|s| {
        for start in (0..n).step_by(chunk) {
            let end = (start + chunk).min(n);
            let f = &f;
            s.spawn(move || f(start, end));
        }
    });
}

impl CpuBackend {
    pub fn new() -> Self {
        Self
    }

    fn ew(&self, a: &CpuTensor, f: impl Fn(f32) -> f32) -> CpuTensor {
        t(a.data.iter().map(|&v| f(v)).collect(), &a.shape)
    }
}

impl Backend for CpuBackend {
    type Tensor = CpuTensor;

    fn name(&self) -> &str {
        "cpu"
    }

    fn tensor(&self, data: &[f32], shape: &[usize]) -> CpuTensor {
        t(data.to_vec(), shape)
    }

    fn quant_tensor(&self, q: QuantWeight) -> CpuTensor {
        let shape = vec![q.shape[0], q.shape[1]];
        CpuTensor { data: Arc::new(Vec::new()), shape, quant: Some(Arc::new(q)) }
    }

    fn zeros(&self, shape: &[usize]) -> CpuTensor {
        t(vec![0.0; shape.iter().product()], shape)
    }

    fn reshape(&self, x: &CpuTensor, shape: &[usize]) -> CpuTensor {
        CpuTensor { data: x.data.clone(), shape: shape.to_vec(), quant: x.quant.clone() }
    }

    fn embed_rows(&self, table: &CpuTensor, ids: &[u32], scale: f32) -> CpuTensor {
        let d = table.shape[1];
        let mut out = vec![0.0f32; ids.len() * d];
        if let Some(q) = &table.quant {
            for (i, &id) in ids.iter().enumerate() {
                for dd in 0..d {
                    out[i * d + dd] = q.deq(id as usize, dd) * scale;
                }
            }
            return t(out, &[ids.len(), d]);
        }
        for (i, &id) in ids.iter().enumerate() {
            let src = id as usize * d;
            for dd in 0..d {
                out[i * d + dd] = table.data[src + dd] * scale;
            }
        }
        t(out, &[ids.len(), d])
    }

    // RMSNorm: y = x/rms(x) * weight. The qat-mobile checkpoint stores norm
    // weights as the FULL multiplier (plain `w`, not HF-Gemma's `1 + w`).
    fn rms_norm(&self, x: &CpuTensor, w: Option<&CpuTensor>, eps: f32) -> CpuTensor {
        let d = *x.shape.last().unwrap();
        let rows = x.size() / d;
        let mut out = vec![0.0f32; rows * d];
        let wd = w.map(|w| w.data.as_slice());
        for (i, o) in out.chunks_mut(d).enumerate() {
            let xr = &x.data[i * d..(i + 1) * d];
            let ss: f32 = xr.iter().map(|v| v * v).sum();
            let inv = 1.0 / (ss / d as f32 + eps).sqrt();
            match wd {
                Some(wd) => {
                    for dd in 0..d {
                        o[dd] = xr[dd] * inv * wd[dd];
                    }
                }
                None => {
                    for dd in 0..d {
                        o[dd] = xr[dd] * inv;
                    }
                }
            }
        }
        t(out, &x.shape)
    }

    fn linear(&self, x: &CpuTensor, w: &CpuTensor) -> CpuTensor {
        let (tt, k) = (x.shape[0], x.shape[1]);
        let n = w.shape[0];
        // Output rows (weight rows) are split across cores; each thread fills
        // its own [n-chunk, T] block of the transposed result, which is then
        // laid out as [T, N]. Per-element products and summation order match
        // the naive loop exactly.
        let mut out_t = vec![0.0f32; n * tt];
        let quant = w.quant.as_deref();
        {
            let out_ptr = SyncPtr(out_t.as_mut_ptr());
            par_ranges(n, n * k * tt, |start, end| {
                // SAFETY: each range writes only rows [start, end) of out_t,
                // and ranges handed to different threads are disjoint.
                let dst = unsafe { std::slice::from_raw_parts_mut(out_ptr.get().add(start * tt), (end - start) * tt) };
                let mut row = vec![0.0f32; k];
                for nn in start..end {
                    let wb: &[f32] = match quant {
                        Some(q) => {
                            // dequantize the weight row once, reuse for every input row
                            for kk in 0..k {
                                row[kk] = q.deq(nn, kk);
                            }
                            &row
                        }
                        None => &w.data[nn * k..(nn + 1) * k],
                    };
                    for i in 0..tt {
                        let xb = &x.data[i * k..(i + 1) * k];
                        let mut acc = 0.0f32;
                        for kk in 0..k {
                            acc += xb[kk] * wb[kk];
                        }
                        dst[(nn - start) * tt + i] = acc;
                    }
                }
            });
        }
        if tt == 1 {
            return t(out_t, &[1, n]);
        }
        let mut out = vec![0.0f32; tt * n];
        for nn in 0..n {
            for i in 0..tt {
                out[i * n + nn] = out_t[nn * tt + i];
            }
        }
        t(out, &[tt, n])
    }

    fn add(&self, a: &CpuTensor, b: &CpuTensor) -> CpuTensor {
        t(a.data.iter().zip(b.data.iter()).map(|(x, y)| x + y).collect(), &a.shape)
    }
    fn mul(&self, a: &CpuTensor, b: &CpuTensor) -> CpuTensor {
        t(a.data.iter().zip(b.data.iter()).map(|(x, y)| x * y).collect(), &a.shape)
    }
    fn scale(&self, a: &CpuTensor, s: f32) -> CpuTensor {
        self.ew(a, |v| v * s)
    }
    fn scale_by_tensor(&self, a: &CpuTensor, s: &CpuTensor) -> CpuTensor {
        self.scale(a, s.data[0])
    }
    fn gelu_tanh(&self, a: &CpuTensor) -> CpuTensor {
        self.ew(a, gelu_tanh_scalar)
    }
    fn geglu(&self, gate: &CpuTensor, up: &CpuTensor) -> CpuTensor {
        self.mul(&self.gelu_tanh(gate), up)
    }
    fn softcap(&self, a: &CpuTensor, cap: f32) -> CpuTensor {
        self.ew(a, |v| cap * (v / cap).tanh())
    }

    // RoPE on q/k laid out [T, heads, headDim]. Half-split convention (HF Gemma),
    // rotating only the first `rotary_dim` dims (partial rotary).
    fn rope(&self, x: &CpuTensor, positions: &[u32], theta: f32, head_dim: usize, heads: usize, rotary_dim: usize) -> CpuTensor {
        let tt = x.shape[0];
        let mut out = x.data.as_ref().clone();
        let half = rotary_dim / 2;
        for i in 0..tt {
            let pos = positions[i] as f32;
            for h in 0..heads {
                let base = (i * heads + h) * head_dim;
                for j in 0..half {
                    let freq = theta.powf(-(2.0 * j as f32) / rotary_dim as f32);
                    let ang = pos * freq;
                    let (sin, cos) = ang.sin_cos();
                    let a = x.data[base + j];
                    let b = x.data[base + j + half];
                    out[base + j] = a * cos - b * sin;
                    out[base + j + half] = b * cos + a * sin;
                }
            }
        }
        t(out, &x.shape)
    }

    // Multi/grouped-query causal attention with optional sliding window.
    // (query, head) rows are independent and split across cores.
    fn attention(&self, q: &CpuTensor, k: &CpuTensor, v: &CpuTensor, o: &AttnOpts) -> CpuTensor {
        let (tt, hq, dh) = (q.shape[0], q.shape[1], q.shape[2]);
        let s = k.shape[0];
        let hkv = k.shape[1];
        let dv = v.shape[2];
        let group = hq / hkv;
        let mut out = vec![0.0f32; tt * hq * dv];
        let out_ptr = SyncPtr(out.as_mut_ptr());
        par_ranges(tt * hq, tt * hq * s * (dh + dv), |start, end| {
            // SAFETY: row r of `out` is written only by the range containing r.
            let dst = unsafe { std::slice::from_raw_parts_mut(out_ptr.get().add(start * dv), (end - start) * dv) };
            let mut scores = vec![0.0f32; s];
            for r in start..end {
                let (i, h) = (r / hq, r % hq);
                let qpos = o.q_pos[i];
                let kvh = h / group;
                let qb = (i * hq + h) * dh;
                let mut maxv = f32::NEG_INFINITY;
                for j in 0..s {
                    let kpos = o.k_pos[j];
                    // causal + optional sliding-window mask
                    if kpos > qpos || (o.sliding_window > 0 && qpos - kpos >= o.sliding_window) {
                        scores[j] = f32::NEG_INFINITY;
                        continue;
                    }
                    let kb = (j * hkv + kvh) * dh;
                    let mut dot = 0.0f32;
                    for d in 0..dh {
                        dot += q.data[qb + d] * k.data[kb + d];
                    }
                    dot *= o.scale;
                    if o.attn_softcap > 0.0 {
                        dot = o.attn_softcap * (dot / o.attn_softcap).tanh();
                    }
                    scores[j] = dot;
                    if dot > maxv {
                        maxv = dot;
                    }
                }
                let mut sum = 0.0f32;
                for j in 0..s {
                    if scores[j] == f32::NEG_INFINITY {
                        scores[j] = 0.0;
                    } else {
                        let e = (scores[j] - maxv).exp();
                        scores[j] = e;
                        sum += e;
                    }
                }
                let inv = if sum > 0.0 { 1.0 / sum } else { 0.0 };
                let ob = (r - start) * dv;
                for j in 0..s {
                    let w = scores[j] * inv;
                    if w == 0.0 {
                        continue;
                    }
                    let vb = (j * hkv + kvh) * dv;
                    for d in 0..dv {
                        dst[ob + d] += w * v.data[vb + d];
                    }
                }
            }
        });
        t(out, &[tt, hq * dv])
    }

    fn slice_rows(&self, x: &CpuTensor, start: usize, count: usize) -> CpuTensor {
        let dim = x.shape[1];
        t(x.data[start * dim..(start + count) * dim].to_vec(), &[count, dim])
    }

    fn slice_cols(&self, x: &CpuTensor, offset: usize, width: usize) -> CpuTensor {
        let (rows, stride) = (x.shape[0], x.shape[1]);
        let mut out = Vec::with_capacity(rows * width);
        for r in 0..rows {
            out.extend_from_slice(&x.data[r * stride + offset..r * stride + offset + width]);
        }
        t(out, &[rows, width])
    }

    fn concat_rows(&self, a: Option<&CpuTensor>, b: &CpuTensor) -> CpuTensor {
        let Some(a) = a else { return t(b.data.as_ref().clone(), &b.shape) };
        let mut out = Vec::with_capacity(a.data.len() + b.data.len());
        out.extend_from_slice(&a.data);
        out.extend_from_slice(&b.data);
        let mut shape = a.shape.clone();
        shape[0] += b.shape[0];
        t(out, &shape)
    }

    fn kv_append(&self, existing: Option<&CpuTensor>, add: &CpuTensor) -> KvAppend<CpuTensor> {
        KvAppend { tensor: self.concat_rows(existing, add), dead: existing.cloned() }
    }

    fn readback(&self, x: &CpuTensor) -> Vec<f32> {
        x.data.as_ref().clone()
    }
}
